### PR TITLE
SITES-17924: Custom AEM Content Fragment Console Action Bar Button

### PR DESCRIPTION
**Pull request by AmyJZhao**
## Description

Add button "Display Metadata" to action bar in AEM CF console that when clicked, displays a modal containing metadata for the selected content fragments

Sites Internal org:
https://experience.adobe.com/?repo=author-p7452-e12437.adobeaemcloud.com&ext=https://245265-601peachweasel.adobeio-static.net/index.html#/@sitesinternal/aem/cf/admin/?appId=aem-cf-admin

Devx Acceleration Prod org:
https://experience.adobe.com/?repo=author-p18453-e47228.adobeaemcloud.com&ext=https://245265-601peachweasel.adobeio-static.net/index.html#/@devxaccelerationprod/aem/cf/admin/?appId=aem-cf-admin

Screenshot 2024-01-20 at 7 31 19 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cf-console-action-bar-button-sample/extension-manifest.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "cf-console-action-bar-button-sample",
+  "id": "cf-console-action-bar-button-sample",
+  "description": "Adding a button to the action bar in AEM Content Fragment Console",
+  "version": "0.0.1",
+  "actionBarButtons": [
+    {
+      "label": "Click Me",
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/extension-manifest.json:8]: Need to fix the `label` and `id`.

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/index.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      <meta http-equiv="X-UA-Compatible" content="ie=edge">
+      <title>cf-console-action-bar-button-sample</title>
+    </head>
+    <body>
+      <noscript>You need to enable JavaScript to run this app.</noscript>
+      <div id="root"></div>
+      <script src="./src/index.js" async="true" type="module"></script>
+    </body>
+    </html>
+
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/index.html:18]:
```suggestion
    
```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/App.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+ * <license header>
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/App.js:2]: Missing license

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ClickMeModal.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
+/*
+ * <license header>
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ClickMeModal.js:2]: Missing license

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ClickMeModal.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+];
+
+const fieldColumns = [
+  {name: 'Name', uid: 'name'},
+  {name: 'Type', uid: 'type'},
+];
+
+const getCfInfoAction = 'get-cf-info';
+
+export default function ClickMeModal () {
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ClickMeModal.js:51]: Please rename the component

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/Constants.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  extensionId: 'cf-console-action-bar-button-sample'
+}
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/Constants.js:3]:
```suggestion
}

```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+/*
+ * <license header>
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js:2]: Missing licence

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  const init = async () => {
+    const guestConnection = await register({
+      id: extensionId,
+      methods: {
+        actionBar: {
+          getButtons() {
+            return [
+              // YOUR ACTION BAR BUTTONS CODE SHOULD BE HERE
+              {
+                'id': 'click-me',
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js:20]: Need to fix the `id`

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/exc-runtime.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/*
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/exc-runtime.js:1]: Is this file needed?

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.css [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.css:1]: Is this file needed?

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  const mockIms = {}
+
+  // render the actual react application and pass along the runtime object to make it available to the App
+  ReactDOM.render(
+    <App runtime={mockRuntime} ims={mockIms} />,
+    document.getElementById('root')
+  )
+}
+
+function bootstrapInExcShell () {
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.js:39]: This function seems redundand.

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please see my comments:
- remove `click-me` references
- remove code that is not used
- add proper license header to each file
-

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                'label': 'Display Metadata',
+                'icon': 'SelectionChecked',
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/ExtensionRegistration.js:22]:
```suggestion
                'icon': 'Info',
```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.css (hunk 10 above)
Comment (AmyJZhao) [reply cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/index.css:1]: I removed the unnecessary CSS, but yes, this makes the accordion item header text bigger and bolder. otherwise, I found it extremely small, light, and hard to read.

File: cf-console-action-bar-button-sample/README.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
+# Custom AEM Content Fragment Console Action Bar Button
+
+Welcome to my Adobe I/O Application!
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/README.md:3]:
```suggestion
A sample extension showcasing adding a button to the Action Bar in AEM Content Fragment Console.
```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/actions/utils.js [hunk 14]
```diff
@@ -0,0 +1,145 @@
+/*
+* <license header>
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/actions/utils.js:2]: Missing license

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/index.html [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  <title>CF Console Action Bar Button</title>
+</head>
+
+<body>
+  <noscript>You need to enable JavaScript to run this app.</noscript>
+  <div id="root"></div>
+  <script src="./src/index.js" async="true" type="module"></script>
+</body>
+
+</html>
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/index.html:29]:
```suggestion
</html>

```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js [hunk 16]
```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2023 Adobe. All rights reserved.
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js:2]:
```suggestion
Copyright 2024 Adobe. All rights reserved.
```

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+        setContentLoading(false);
+      }
+    })()
+  }, [])
+
+  const onCloseHandler = () => {
+    guestConnection.host.modal.close()
+  }
+
+  return (isContentLoading || !guestConnection) ? (
```
Comment (lenaorobei) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js:91]: Looks like this file needs to be formatted

**Review by slitviachenko — CHANGES_REQUESTED** [review]
@AmyJZhao , please apply this workaround, https://github.com/adobe/aem-uix-examples/pull/14 for fixing an issue with a right scroll bar
Screenshot 2024-01-23 at 9 41 41 AM

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+  // Fields
+  const [isContentLoading, setContentLoading] = useState(true);
+  const [guestConnection, setGuestConnection] = useState();
+  const [error, setError] = useState(null);
+  const [fragments, setFragments] = useState([]);
+  const { selection } = useParams();
+  const fragmentIds = selection?.split('|') || [];
+
+  console.log('Selected Fragment Ids', fragmentIds);
+  if (!fragmentIds || fragmentIds.length === 0) {
```
Comment (slitviachenko) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js:60]:
```suggestion
  if (fragmentIds.length === 0) {
```
from your code `fragmentIds` will be always an array

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+];
+
+const fieldColumns = [
+  {name: 'Name', uid: 'name'},
+  {name: 'Type', uid: 'type'},
+];
+
+export default function DisplayMetadataModal () {
+  // Fields
+  const [isContentLoading, setContentLoading] = useState(true);
```
Comment (slitviachenko) [inline cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js:52]:
```suggestion
  const [isContentLoading, setIsContentLoading] = useState(true);
```
to be more consistent with React established practice

File: cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js (hunk 17 above)
Comment (AmyJZhao) [reply cf-console-action-bar-button-sample/src/aem-cf-console-admin-1/web-src/src/components/DisplayMetadataModal.js:91]: @lenaorobei can you elaborate on this?

**Review by lenaorobei — APPROVED** [review]

**Comment by AmyJZhao** [comment]
Known scrollbar issue
